--- uboot/Makefile	2020-02-06 11:40:25.543766424 +0800
+++ uboot_new/Makefile	2020-02-06 11:40:23.439818608 +0800
@@ -852,6 +852,10 @@
 ALL-y += u-boot-sunxi-with-spl.bin
 endif
 
+#
+ifeq ($(CONFIG_ARCH_ASPEED)$(CONFIG_SPL),yy)
+ALL-y +=simage.bin
+endif
 # enable combined SPL/u-boot/dtb rules for tegra
 ifeq ($(CONFIG_TEGRA)$(CONFIG_SPL),yy)
 ALL-y += u-boot-tegra.bin u-boot-nodtb-tegra.bin
@@ -1387,6 +1391,10 @@
 endif  # binman
 endif
 
+OBJCOPYFLAGS_simage.bin = -O binary --pad-to=$(CONFIG_SYS_TEXT_BASE)
+simage.bin: spl/u-boot-spl u-boot.bin FORCE
+	$(call if_changed,pad_cat)
+
 OBJCOPYFLAGS_u-boot-app.efi := $(OBJCOPYFLAGS_EFI)
 u-boot-app.efi: u-boot FORCE
 	$(call if_changed,zobjcopy)
