diff -Naur uboot/drivers/clk/aspeed/clk_ast2600.c uboot_new/drivers/clk/aspeed/clk_ast2600.c
--- uboot/drivers/clk/aspeed/clk_ast2600.c	2020-07-20 13:40:27.221243232 +0800
+++ uboot_new/drivers/clk/aspeed/clk_ast2600.c	2020-07-20 13:40:41.492499314 +0800
@@ -18,9 +18,9 @@
 #define RGMII_TXCLK_ODLY	8
 #define RMII_RXCLK_IDLY		2
 
-#define MAC_DEF_DELAY_1G	0x00410410
-#define MAC_DEF_DELAY_100M	0x00410410
-#define MAC_DEF_DELAY_10M	0x00410410
+#define MAC_DEF_DELAY_1G	0x0041b410
+#define MAC_DEF_DELAY_100M	0x00417410
+#define MAC_DEF_DELAY_10M	0x00417410
 
 #define MAC34_DEF_DELAY_1G	0x00104208
 #define MAC34_DEF_DELAY_100M	0x00104208
